cornflow-server/cornflow/shared/authentication/auth.py

```diff
@@ -195,65 +187,175 @@ def get_token_from_header(headers: Headers = None) -> str:
-            err = "User does not exist, invalid token."
-            raise ObjectDoesNotExist(
-                err, log_txt="Error while trying to get user from header. " + err
+            err = "User not found. Please ensure you are using valid credentials"
+            raise InvalidCredentials(
+                err, 
+                log_txt="Error while trying to get user from header. User does not exist.",
+                status_code=400
             )
         return user
 
     @staticmethod
-    def return_user_from_token(token):
+    def get_public_keys(provider_url: str) -> dict:
         """
-        Function used for internal testing. Given a token gives back the user_id encoded in it.
+        Gets and caches the public keys from the OIDC provider
+        
+        :param str provider_url: The base URL of the OIDC provider
+        :return: Dictionary of kid to public key mappings
+        :rtype: dict
+        """
+        # Check cache first
+        if provider_url in public_keys_cache:
+            return public_keys_cache[provider_url]
+            
+        # Fetch keys from provider
+        jwks_url = f"{provider_url.rstrip('/')}/.well-known/jwks.json"
+        try:
+            response = requests.get(jwks_url)
+            response.raise_for_status()
+            
+            # Convert JWK to RSA public keys using PyJWT's built-in method
+            public_keys = {
+                key["kid"]: RSAAlgorithm.from_jwk(key) 
+                for key in response.json()["keys"]
+            }
+            
+            # Store in cache
+            public_keys_cache[provider_url] = public_keys
+            return public_keys
+            
+        except requests.exceptions.RequestException as e:
+            raise CommunicationError(
+                "Failed to fetch public keys from authentication provider",
+                log_txt=f"Error while fetching public keys from {jwks_url}: {str(e)}",
+                status_code=400
+            )
 
-        :param str token: the given token
-        :return: the user id code.
-        :rtype: int
+    def verify_token(self, token: str, provider_url: str, expected_audience: str) -> dict:
+        """
+        Verifies an OpenID Connect token
+        
+        :param str token: The token to verify
+        :param str provider_url: The base URL of the OIDC provider
+        :param str expected_audience: The expected audience claim
+        :return: The decoded token claims
+        :rtype: dict
         """
-        user_id = Auth.decode_token(token)["user_id"]
-        return user_id
+        # First validate token header
+        try:
+            unverified_header = jwt.get_unverified_header(token)
+        except jwt.InvalidTokenError:
+            raise InvalidCredentials(
+                "Invalid token format or signature",
+                log_txt="Error while verifying token. Invalid token format.",
+                status_code=400
+            )
```

Comment:
This is what i meant, this except already exists on the outer scope of this method. So lets try to have just one